discord/rate_limiting.py

```diff
@@ -0,0 +1,562 @@
+                    self.release(1)
+
+                raise
+            else:
+                if self.remaining is not None:
+                    self.release(self.remaining)
+                elif self.metadata_unknown:
+                    self.release()
+                else:
+                    if self._pending.qsize() >= 1:
+                        self.release(1)
+            finally:
+                self._reserved.remove(fut)
+                self._fetch_metadata.set()
+
+            return
+
+        await self._fetch_metadata.wait()
+
+        async with self.reserve(priority=priority):
+            yield
+
+    def release(self, count: int | None = None) -> None:
+        """Release *count* amount of requests.
+
+        .. warning:: This should not be used directly.
+
+        Parameters
+        ----------
+        count: :class:`int`
+            The number of requests to release.
+        """
+
+        if count is not None:
+            num = min(count, self._pending.qsize())
+        else:
+            num = self._pending.qsize()
+
+        for _ in range(num):
+            fut = self._pending.get_nowait()
+
+            self._pending.task_done()
+
+            fut.set_result(None)
+
+    @property
+    def garbage(self) -> bool:
+        """Whether this bucket should be collected by the garbage collector."""
+
+        # the bucket should be collected if it has no limit and is not reserved
+        return self.metadata_unknown and not self._reserved
+
+    async def stop(self) -> None:
+        """Cancel all reserved futures from use."""
+
+        for fut in self._reserved:
+            fut.set_exception(asyncio.CancelledError)
+
+    def set_metadata(
+        self,
+        remaining: int | None = None,
+        reset_after: int | None = None,
+    ) -> None:
+        """Set the metadata for this Bucket.
+        If both `remaining` and `reset_after` are set to ``None`` and
+        :attr:`metadata_unknown` is set to ``True``, a garbage state will
+        be initialized and all pending requests will be released.
+
+        Parameters
+        ----------
+        remaining: Optional[:class:`int`]
+            The remaining number of requests which can be made
+            to this bucket.
+        reset_after: Optional[:class:`int`]
+            The number of seconds until this bucket will next reset.
+        """
+
+        if remaining is None and reset_after is None:
+            self.metadata_unknown = True
+            self.release()
+        else:
+            self.remaining = remaining
+
+            if self._reset_after_set:
+                return
+
+            self._reset_after_set = True
+
+            loop = asyncio.get_running_loop()
+            loop.call_later(cast(float, reset_after), self._reset)
+
+    def _reset(self) -> None:
+        self._reset_after_set = False
+        self.remaining = None
+
+        self.release(self.limit)
+
+
+class BucketStorageProtocol(Protocol):
```

Comment:
I would recommend making this an ABC instead since protocols cannot be checked for at runtime unless you use a decorator.